### ISSUE TITLE
aggregate / transform hazard to geometry of exposure before calculating impact?!

### ISSUE THREAD

**Issue by togeig** [CLIMADA-project/climada_python#13]
Hi all,
I am struggling to find the best CLIMADA-thonic way of implementing following situation:

I have a hazard on a regular raster and exposure available on NUTS3-level. Instead of remapping my exposure to the hazard, I want to statistically aggregate (e.g. max, mean value) the hazard across corresponding NUTS3 regions and then match with the exposure to determine the impact. 

I understand that CLIMADA usually works the opposite way, as everything is adjusted to the hazard. However, I find this not very practical in my case as the impact needs to be defined at NUTS3-level in the end. Any suggestions of how to deal with that?

The optimal way in my eyes would look like that:
1) create a geodataframe from the hazard and use the dissolve function to aggregate the hazard to NUTS3 region. Then calculate the impact based on the NUTS3 regions only.

My (not optimal) ideas for a workaround are these:
1) preprocess hazard data (aggregate to NUTS3 regions) and continue with CLIMADA from that point.
2) create downscaled raster exposure by just populating grid points within a NUTS3 region with identical exposure value. Once the impact is then calculated I have to postprocess this data to aggregate the result to NUTS3 level.

Any suggestions on how to deal with this within CLIMADA greatly appreciated. If nothing like this is presently implemented, happy to discuss pros and cons of making it possible.

Thanks, Tobias

**Comment by ThomasRoosli** [comment]
Hi Tobias
That is an interesting question, I look forward to the discussion and I answer with my first thoughts. 

Your proposed optimal way 1. and not optimal idea 1. sound very similar to me. There is a preprocessing step to aggregate the hazard from grid to the NUTS3 region. This is certainly a possibility and CLIMADA has some traces of such approaches (e.g., the attribute "fraction" in the Hazard class would allow you to save one intensity per NUTS3 region and then specify a certain fraction of the NUTS3 exposure which would experience this intensity. This might be useful for a flood hazard, for example.

Personally, my view of the CLIMADA-thonic way would be to stay at the highest resolution possible.  The question is, can the exposure be disaggregated reasonably to a higher solution, but in a better approach than just assuming identical exposure values. This depends on the type of exposure in focus in your study.
1. one could think of disaggregation of the exposure value per NUTS3 region by a gridded population dataset (look at the LitPop methodology for input on how to implement such a disaggregation).  
2. Or disagregate evenly over some part of the area (high value areas), and not over other parts (low value area like lakes, forests etc.) See the Use Case 1 in this tutorial (still under development; https://github.com/CLIMADA-project/climada_python/blob/master/doc/tutorial/climada_entity_openstreetmap.ipynb). Instead of Open Street Map, different Datasets like Corine landcover could be used.

I am interested to hear the suggestions of others.

**Comment by mmyrte** [comment]
I just found a graph that better describes my idea:
![image](https://user-images.githubusercontent.com/24587121/78604030-6929f000-7859-11ea-8a88-eb7972f67547.png)

Which is sourced from the [pandarus docs](https://pandarus.readthedocs.io/).

**Comment by togeig** [comment]
Dear Jan and Thomas,
thanks for your feedback...very appreciated! Let me respond with a few comments:

@ThomasRoosli : I Have also been thinking about your suggestion 1) and will probably try to work in this direction. I favor the high-res approach as well but not very clear to how I should proceed after calculating the high-res impact when wanting to aggregate the final result to NUTS3 level? Any direct CLIMADA functionality available for this currently? I guess I have to transfer it to an Exposure again to be able to use the geodataframe functionality. 
Suggestion 2) sounds also interesting, I will have closer look soon. 

@mmyrte : Your suggestion sounds like a feature that CLIMADA should offer. Maybe the remapping function in the impact calculation (i.e. in interpolation.py in util) can be advanced to offer a remapping including weighting and/or selecting. I my case the hazard is very high-res and each exposure centroid could coincide with hundreds of hazard points. Choosing only the nearest point can be very misleading. So being able to select at least the max or mean values from a specific radius (or ideally the full NUTS3 region) would help.

**Comment by gabrielaznar** [comment]
Hi Tobias,

that's a good point and worth to take into account as a possible use case. Currently it is left to the user how to handle the different resolution and geometry issues between exposure and hazard in a "preprocessing" step. climada only makes some checks internally and offers some functionality to transform. It is not all adjusted to the hazard. In the tutorial on end-to-end impact calculations ([https://climada-python.readthedocs.io/en/latest/tutorial/climada_engine_Impact.html]) it is explained how to handle point and raster exposures. Of course, the third natural candidate are polygon exposures and I wanted to have had implemented that as well.

The idea is that the polygon of each exposure can be defined in the geometry column of the exposure geodataframe, and lat/lon would make reference to the centroids of the polygon. Then, the idea was to implement the functionality of aggregating raster cells to polygon shapes according to different criteria (mean, max, min, weighted according to area as proposed by @mmyrte etc) somewhere and that the assign_centroids method of the exposure could somehow connect to that. I'd love to implement something like that. If you implement some of that we could discuss where to put it. I'd be happy to discuss further.

Best,
Gabriela

**Comment by ThomasRoosli** [comment]
@gabrielaznar it would be great if we could think about how to assign the centroids, do the impact calculation and plotting for polygon exposures, but also for line-segments to represent mobility infrastructure.
@togeig as a pragmatic solution (which I apply for Switzerland): I create a raster exposure and then preprocess it to represent polygons. I use the functionality of util.coordinates.coord_on_land() to find all exposure gridpoints within a polygon. I then populate an additional column in the exposure with this information, this calculation takes a while. If I then use the flag "save_mat=True" in Impact.calc(), I can use the additional column to aggregate Impacts easily and quickly to the polygon regions.

**Comment by togeig** [comment]
Hi all,
thanks for this very useful discussion.

@gabrielaznar : Your idea you wanted to implement sounds exactly of what I'd need. Plus the addition of @ThomasRoosli to add functionality for line polygons. Happy to discuss this further, maybe we need to have a videocall on that one.

@ThomasRoosli : thanks for your tips on the workaround. How do you create your raster exposure, e.g. using LitPop based on your polygon data? And you simply save a list of raster points as a cell in the geodateframe? Could you share a code sample for your workaround? If your not interested on sub-polygon resolution, you could probably also simply use your polygon exposure, create a regular raster based on the bounding box of your polygon and find the raster points within the polygon, e.g. using coord_on_land() or similar functions and then proceed as suggest by you.

**Comment by chahank** [comment]
@togeig Is this issue still open?

**Comment by togeig** [comment]
Yes, i haven't had the time and the others neither. We were discussing using the rasterstats package for that but that's it so far. 

Am 18. September 2020 15:54:08 MESZ schrieb Chahan Kropf <notifications@github.com>:

**Comment by chahank** [comment]
This is being addressed in PR #351, although with a different approach. The lines/polygons handling is on the level of the `exposures`. Thus, the NUTs polygons can be disaggregated on a grid, then the impact is computed on this grid, and finally the impact is aggregated by polygon. If time permits, it would be good to receive feedback from @togeig @ThomasRoosli @mmyrte @gabrielaznar  whether the PR fulfills the needs from this issue. There is a tutorial in the PR #351 which however is not yet complete. The branch is work in progress, but with a already solid backbone.

**Comment by mmyrte** [comment]
I'm afraid I don't have time/value my free weekend a bit too highly right now, so **big caveat**: I only skimmed the PR! This may well be bollocks:

I'm not sure how accessible the functionality in the PR is if it's not implemented as a bunch of class methods - it would seem that keeping track of the various aggregated/disaggregated states is ideally suited to an OO approach.

Also tagging @Evelyn-M just in case this is helpful. In general: cool to have the functionality! Thanks a bunch!

**Comment by chahank** [comment]
Thanks @mmyrte for the feedback! Your comment actually intersects well with the current implementation. New Exposures and Impact Objects are made with aggregated/disaggregated information. We are not yet sure how exactly this will be included in the existing classes, thus it is currently written as a bunch of util methods.
